File: rest/app/controllers/gateway.py
from fastapi import APIRouter, Depends, HTTPException, Request
from app.services.user import UserService
from app.core.security import create_access_token
from app.core.consul import get_service_instance, NoRecordFound
from app.core.config import settings
from datetime import timedelta

router = APIRouter()


@router.post("/get")
def get_gateway_credentials(request: Request, user_service: UserService = Depends()):
    user = user_service.check_auth(request)
    gateway_token = create_access_token(
        str(user.id), timedelta(minutes=1), secret_key=settings.GATEWAY_SECRET_KEY
    )
    try:
        gateway_ip = get_service_instance("GATEWAY")
    except NoRecordFound:
        raise HTTPException(status_code=503, detail="Gateway node not available")
    # The access token will be used by the client to authenticate with gateway
    return {"token": gateway_token, "endpoint": gateway_ip}

File: web/src/http/gateway.ts
import { client } from ".";
import { ENDPOINTS } from "./endpoints";

export interface GatewayCredentials {
  token: string;
  endpoint: string;
}

export const getGatewayCredentials = async () => {
  const resp = await client.post(ENDPOINTS.GATEWAY.GET_GATEWAY_CREDENTIALS);
  return resp.data as GatewayCredentials;
};
